src/EventStore.Core/Services/Transport/Tcp/TcpConnectionManager.cs

```diff
@@ -410,30 +410,48 @@ public class TcpConnectionManager : IHandle<TcpMessage.Heartbeat>, IHandle<TcpMe
 		public void Handle(TcpMessage.Heartbeat message) {
 			if (IsClosed) return;
 
-			var msgNum = _messageNumber;
-			if (message.MessageNumber != msgNum)
-				ScheduleHeartbeat(msgNum);
-			else {
-				SendPackage(new TcpPackage(TcpCommand.HeartbeatRequestCommand, Guid.NewGuid(), null));
+			var receiveProgressIndicator = _receiveProgressIndicator;
+			var sendProgressIndicator = _sendProgressIndicator;
 
+			/*
+			 * If we have not received any data within the heartbeat interval and we are not waiting for a heartbeat timeout check, then we send a heartbeat request.
+			 */
+			if (message.ReceiveProgressIndicator == receiveProgressIndicator && !_awaitingHeartbeatTimeoutCheck) {
+				SendPackage(new TcpPackage(TcpCommand.HeartbeatRequestCommand, Guid.NewGuid(), null));
+				_awaitingHeartbeatTimeoutCheck = true;
 				_publisher.Publish(TimerMessage.Schedule.Create(_heartbeatTimeout, _weakThisEnvelope,
-					new TcpMessage.HeartbeatTimeout(msgNum)));
+					new TcpMessage.HeartbeatTimeout(receiveProgressIndicator)));
 			}
+			/*
+			 * As a proactive measure, if we have not sent any data to the remote party within the heartbeat interval,
+			 * we also send a heartbeat request just to generate some data for the remote party so that it can clear its heartbeat timeouts.
+			 * This is particularly useful when the remote party's heartbeat request is stuck behind a large (multi-megabyte) TCP message.
+			 */
+			else if (message.SendProgressIndicator == sendProgressIndicator) {
+				SendPackage(new TcpPackage(TcpCommand.HeartbeatRequestCommand, Guid.NewGuid(), null));
+				Log.Trace(
```

Comment:
Trace logs are enabled by default in v5 which means that the logs for an idle cluster may fill up with this message.
Is this log necessary?

Reply:
I've commented out the Log.Trace line, thanks @hayley-jean